src/concurrency/init_once.rs

```diff
@@ -172,26 +165,17 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
         );
 
         // Each complete happens-before the end of the wait
-        // FIXME: should this really induce synchronization? If we think of it as a lock, then yes,
-        // but the docs don't talk about such details.
         if let Some(data_race) = &this.machine.data_race {
             data_race.validate_lock_release(&mut init_once.data_race, current_thread);
         }
 
         // Wake up one waiting thread, so they can go ahead and try to init this.
         if let Some(waiter) = init_once.waiters.pop_front() {
-            // End of the wait happens-before woken-up thread.
-            if let Some(data_race) = &this.machine.data_race {
-                data_race.validate_lock_acquire(
-                    &this.machine.threads.sync.init_onces[id].data_race,
-                    waiter.thread,
-                );
-            }
-
             this.unblock_thread(waiter.thread);
 
             // Call callback, with the woken-up thread as `current`.
             this.set_active_thread(waiter.thread);
+            this.init_once_acquire(id);
```

Comment:
This is a strange one, since we didn't actually fully init the init_once yet. It's not really documented to provide any synchronization in that case, is it?

Reply:
Well, synchronization is not documented in any case. If there wasn't synchronization here, you could get a data race between a failed initialization attempt and the next initialization attempt, which doesn't make any sense.

Reply:
If InitOnce used relaxed accesses plus fences only for the success case, this could totally happen. So it makes sense to me.

But I agree that is unlikely to be the implementation.

> Well, synchronization is not documented in any case.

Yeah and that's a problem, concurrent functions should *always* document their synchronization -- if they don't document any synchronization they should be assumed not to synchronize. This is an important part of the observable behavior of a function.